src/sudoku.py

```diff
@@ -68,21 +79,124 @@ def set_cell(self, x: int, y: int, symbol: int) -> bool:
+
+    def has_unique_solution(self) -> bool:
+        """returns True if the grid has one unique solution, False otherwise"""
+        return self.__has_solutions() == 1
+
+    def __has_solutions(self, x=0, y=0, count=0) -> int:
+        """Recursive backtracking algorithm.
+        Returns:
+            '0' if no solution was found
+            '1' if the solution is unique
+            more than '1' if there are multiple solutions
+        """
+
+        # if (x, y) is outside the grid, then a solution was found
+        if not self.is_cell_inside(x, y):
+            return count + 1
+
+        next_x, next_y = self.__get_next_cell_coordinates(x, y)
+
+        # if the cell is fixed, skip it
+        if self.fixed[x][y]:
+            return self.__has_solutions(next_x, next_y, count)
+
+        old_symbol = self.cells[x][y]
+
+        # try to place every symbol in the cell
+        for symbol in self.get_symbols():
+            if self.set_cell(x, y, symbol):
+                count = self.__has_solutions(next_x, next_y, count)
+                self.cells[x][y] = old_symbol
+
+                # multiple solutions were found: stop searching for more
+                if count > 1:
+                    break
+        return count
+
+    def generate_grid(self, iterations: int) -> int:
+        self.__fill_grid(0, 0)
+
+        count = 0
+        for i in range(iterations):
+            print('iteration: ', i)
+            if self.__try_to_clear_random_cell():
+                count += 1
+        return count
+
+    def __fill_grid(self, x: int, y: int) -> bool:
+        """randomly fill a grid with fixed values"""
+
+        # if (x, y) is outside the grid, then the grid is entirely full
+        if not self.is_cell_inside(x, y):
+            return True
+
+        next_x, next_y = self.__get_next_cell_coordinates(x, y)
+        symbols = random.sample(self.get_symbols(), k=self.size)
+
+        # try to place every symbol in the cell
+        for symbol in symbols:
+            if self.set_fixed(x, y, symbol):
+                if self.__fill_grid(next_x, next_y):
+                    return True # the grid is full
+
+                # this symbol didn't work: go back
+                self.cells[x][y] = None
+                self.fixed[x][y] = False
+
+        # no symbol could be placed
+        return False
+
+    def __try_to_clear_random_cell(self) -> bool:
+        """attempt to remove one symbol from the grid while maintaining
+        uniqueness of the solution"""
+
+        # find (x, y) coordinates of a random fixed cell
+        for y in random.sample(range(self.size), k=self.size):
+            for x in random.sample(range(self.size), k=self.size):
+                if self.fixed[x][y]:
+                    break
+
+        # clear the cell
+        symbol = self.cells[x][y]
+        self.cells[x][y] = None
+        self.fixed[x][y] = False
+
+        # check if the solution is still unique
+        if not self.has_unique_solution():
```

Comment:
Non riesco a capire dove sono definiti x e y 💀

Reply:
Escono fuori dai for loop. Le variabili hanno scope a livello di funzione. Magari non è il modo più elegante per scriverlo.